### PR TITLE
[core] Add metrics for disk and network I/O

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Adds some metrics useful for object-intensive workloads:
- Per raylet/object manager:
  - Add num bytes pending restore to spill manager
  - Add num requests cumulative to PullManager
  - Num bytes pushed/pulled from other nodes cumulative
  - Histogram for request latencies in PullManager:
    -  total life time of request, from start to cancel
    - request satisfaction time, from start to object local
    - pull time, from object activation to object local
- Per-node disk read/write speed, IOPS

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/reporter/reporter_agent.py [hunk 1]
```diff
@@ -348,19 +361,25 @@ def _get_load_avg(self):
         per_cpu_load = tuple((round(x / self._cpu_counts[0], 2) for x in load))
         return load, per_cpu_load
 
+    @staticmethod
+    def _compute_speed_from_hist(hist):
+        while len(hist) > 7:
```
Comment (jjyao) [inline dashboard/modules/reporter/reporter_agent.py:366]: why is 7 here?

File: dashboard/modules/reporter/reporter_agent.py [hunk 2]
```diff
@@ -348,19 +361,25 @@ def _get_load_avg(self):
...
-        then, prev_network_stats = self._network_stats_hist[0]
-        prev_send, prev_recv = prev_network_stats
-        now_send, now_recv = network_stats
-        network_speed_stats = (
-            (now_send - prev_send) / (now - then),
-            (now_recv - prev_recv) / (now - then),
-        )
+        network_speed_stats = self._compute_speed_from_hist(self._network_stats_hist)
+
+        disk_stats = self._get_disk_stats()
```
Comment (jjyao) [inline dashboard/modules/reporter/reporter_agent.py:379]: there is only `_get_disk_io` function?

File: dashboard/modules/reporter/reporter_agent.py (hunk 1 above)
Comment (stephanie-wang) [reply dashboard/modules/reporter/reporter_agent.py:366]: Not sure, I just copied what was here already...

File: dashboard/modules/reporter/reporter_agent.py (hunk 2 above)
Comment (stephanie-wang) [reply dashboard/modules/reporter/reporter_agent.py:379]: Oops, thanks... clearly I did not actually run this :)

Comment (jjyao) [reply dashboard/modules/reporter/reporter_agent.py]: Can we add some tests?

File: src/ray/object_manager/pull_manager.cc [hunk 3]
```diff
@@ -651,10 +660,20 @@ bool PullManager::TryPinObject(const ObjectID &object_id) {
     if (ref != nullptr) {
       pinned_objects_size_ += ref->GetSize();
       pinned_objects_[object_id] = std::move(ref);
-      return false;
```
Comment (scv119) [inline src/ray/object_manager/pull_manager.cc:654]: looks this changes the semantics of TryPinObject, is it intentional?

File: src/ray/object_manager/pull_manager.cc [hunk 4]
```diff
@@ -172,6 +171,10 @@ bool PullManager::ActivateNextPullBundleRequest(const Queue &bundles,
       active_object_pull_requests_[obj_id].insert(next_request_it->first);
       if (needs_pull) {
         RAY_LOG(DEBUG) << "Activating pull for object " << obj_id;
+        auto it = object_pull_requests_.find(obj_id);
+        RAY_CHECK(it != object_pull_requests_.end());
```
Comment (jjyao) [inline src/ray/object_manager/pull_manager.cc:175]: You can use `map_find_or_die()` which gives you better check failure message.

File: src/ray/object_manager/pull_manager.cc [hunk 5]
```diff
@@ -640,10 +649,20 @@ bool PullManager::TryPinObject(const ObjectID &object_id) {
     if (ref != nullptr) {
```
Comment (jjyao) [inline src/ray/object_manager/pull_manager.cc:649]: It's also useful to record num_failed_pins_

File: src/ray/object_manager/pull_manager.cc [hunk 6]
```diff
@@ -640,10 +649,20 @@ bool PullManager::TryPinObject(const ObjectID &object_id) {
...
+
+      auto it = object_pull_requests_.find(object_id);
+      RAY_CHECK(it != object_pull_requests_.end());
+      ray::stats::STATS_pull_manager_object_request_time_ms.Record(
+          absl::GetCurrentTimeNanos() / 1e3 - it->second.request_start_time_ms,
+          "StartToPin");
+      if (it->second.activate_time_ms > 0) {
+        ray::stats::STATS_pull_manager_object_request_time_ms.Record(
+            absl::GetCurrentTimeNanos() / 1e3 - it->second.activate_time_ms,
+            "MemoryAvailableToPin");
```
Comment (jjyao) [inline src/ray/object_manager/pull_manager.cc:661]: nit: ActivateToPin?

**Review by jjyao — APPROVED** [review]

**Comment by stephanie-wang** [comment]
@scv119 @ericl or @rkooo567 could you approve as a codeowner?

**Review by scv119 — APPROVED** [review]

**Pull request merged**